File: src/main/java/antimattaermod/core/Util/BlockPos.java
package antimattaermod.core.Util;

import net.minecraft.block.Block;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;

/**
 * @author C6H2Cl2
 */
public class BlockPos {
    private int x,y,z;
    public BlockPos(int x,int y,int z){
        this.x = x;
        this.y = y;
        this.z = z;
    }
    public BlockPos getUp(){
        return new BlockPos(x,y+1,z);
    }

    public BlockPos getDown(){
        return new BlockPos(x,y-1,z);
    }

    public BlockPos getNorth(){
        return new BlockPos(x,y,z-1);
    }

    public BlockPos getSouth(){
        return new BlockPos(x,y,z+1);
    }

    public BlockPos getEast(){
        return new BlockPos(x+1,y,z);
    }

    public BlockPos getWest(){
        return new BlockPos(x-1,y,z);
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public int getZ() {
        return z;
    }

    public Block getBlockFromPos(World world){
        return world.getBlock(x,y,z);
    }

    public TileEntity getTileEntityFromPos(World world){
        return world.getTileEntity(x,y,z);
    }
}

File: src/main/java/antimattermod/core/Util/BlockPos.kt
package antimattermod.core.Util

import net.minecraft.block.Block
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.IBlockAccess
import net.minecraft.world.IWorldAccess
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection

/**
 * @author C6H2Cl2
 */
data class BlockPos(var x: Int,var y: Int,var z: Int) {

    val up: BlockPos
        get() = BlockPos(x, y + 1, z)

    val down: BlockPos
        get() = BlockPos(x, y - 1, z)

    val north: BlockPos
        get() = BlockPos(x, y, z - 1)

    val south: BlockPos
        get() = BlockPos(x, y, z + 1)

    val east: BlockPos
        get() = BlockPos(x + 1, y, z)

    val west: BlockPos
        get() = BlockPos(x - 1, y, z)

    fun getBlockFromPos(world: World): Block {
        return world.getBlock(x, y, z)
    }

    fun getTileEntityFromPos(world: IBlockAccess): TileEntity {
        return world.getTileEntity(x, y, z)
    }

    fun getBlockDirection(tilePos: BlockPos): ForgeDirection {
        if (up == tilePos) {
            return ForgeDirection.UP
        } else if (down == tilePos) {
            return ForgeDirection.DOWN
        } else if (north == tilePos) {
            return ForgeDirection.NORTH
        } else if (south == tilePos) {
            return ForgeDirection.SOUTH
        } else if (east == tilePos) {
            return ForgeDirection.EAST
        } else if (west == tilePos) {
            return ForgeDirection.WEST
        } else {
            return ForgeDirection.UNKNOWN
        }
    }

    @JvmOverloads fun writeToNBT(tagCompound: NBTTagCompound, tagName: String = "blockPos") {
        val tag = NBTTagCompound()
        tag.setInteger("x", x)
        tag.setInteger("y", y)
        tag.setInteger("z", z)
        tagCompound.setTag(tagName, tag)
    }

    @JvmOverloads fun readFromNBT(tagCompound: NBTTagCompound, tagName: String = "blockPos") {
        val tag = tagCompound.getCompoundTag(tagName)
        x = tag.getInteger("x")
        y = tag.getInteger("y")
        z = tag.getInteger("z")
    }
}
